Update Databricks docs to include IPython and Kedro-viz workaround (#1486)
kedro-org/kedro: docs/source/deployment/databricks.md

Your complete notebook should look similar to this (the results are hidden):

![](../meta/images/databricks_notebook_example.png)

### 9. Using IPython Extension 

Kedro's IPython extension can be used in a Databricks notebook in a similar way to how it is used in [Jupyter notebooks](https://kedro.readthedocs.io/en/0.17.4/11_tools_integration/02_ipython.html#ipython-extension)

> There's one additional installation you need to do in the Databricks notebook to avoid `ContextualVersionConflict` error.

Reviewer: I don't think it's necessary to mention this error here. I'd change it to:
```suggestion
There's one additional installation you need to do in the Databricks notebook to make use of the IPython extension. 
```

Your complete notebook should look similar to this (the results are hidden):

![](../meta/images/databricks_notebook_example.png)

### 9. Using IPython Extension 

You can interact with Kedro in Databricks through the Kedro [IPython extension](https://ipython.readthedocs.io/en/stable/config/extensions/index.html), `kedro.extras.extensions.ipython`.

> The Kedro IPython extension launches a [Kedro session](../kedro_project_setup/session.md) and makes available the useful Kedro variables `catalog`, `context`, `pipelines` and `session`. It also provides the `%reload_kedro`  [line magic](https://ipython.readthedocs.io/en/stable/interactive/magics.html) that reloads these variables (for example, if you need to update `catalog` following changes to your Data Catalog).

Reviewer 2: ```suggestion
The Kedro IPython extension launches a [Kedro session](../kedro_project_setup/session.md) and makes available the useful Kedro variables `catalog`, `context`, `pipelines` and `session`. It also provides the `%reload_kedro` [line magic](https://ipython.readthedocs.io/en/stable/interactive/magics.html) that reloads these variables (for example, if you need to update `catalog` following changes to your Data Catalog).
```

> The IPython extension can be used in a Databricks notebook in a similar way to how it is used in [Jupyter notebooks](https://kedro.readthedocs.io/en/0.18.0/tools_integration/ipython.html)

Reviewer 2: ```suggestion
The IPython extension can be used in a Databricks notebook in a similar way to how it is used in [Jupyter notebooks](https://kedro.readthedocs.io/en/0.18.0/tools_integration/ipython.html).
```

> There's one additional installation you need to do in the Databricks notebook to make use of the IPython extension. After you load the IPython extension using the below command:

Reviewer 2: Could we maybe add why we need to do this workaround for users?

```ipython
In [1]: %load_ext kedro.extras.extensions.ipython
```

> you must explicitly upgrade your pip version by doing the below:

Reviewer 2: ```suggestion
You must explicitly upgrade your `pip` version by doing the below:
```

```bash
%pip install -U pip
```

> after this, you can reload kedro by running the line magic command  `%reload_kedro <path_to_project_root>`.

Reviewer 2: ```suggestion
After this, you can reload Kedro by running the line magic command `%reload_kedro <path_to_project_root>`.
```

> ### 10. Running Kedro-viz on Databricks 

Reviewer 2: ```suggestion
### 10. Running Kedro-Viz on Databricks 
```

Reviewer 2: Remember to capitalise the `v` in `Kedro-Viz`, [according to the style guide](https://kedro.readthedocs.io/en/stable/contribution/documentation_contributor_guidelines.html#kedro-lexicon). I spotted this one a few times but I don't want to leave comments for it.

Currently, if you try to run `%run_viz` on Databricks it will only display the below: 

```console
<IPython.core.display.HTML object>
```

> While we fix this issue, we have a temporary workaround to run Kedro-viz on Databricks. 

Reviewer 2: It would be great to see something explaining why this happens. I'd add in something like, "This problem is caused by ... and while we fix this issue, we have a temporary workaround which involves you setting up a R Shiny application to run Kedro-Viz on Databricks."

Author: we haven't figured out why this problem is caused. it looks like a Databricks issue! so i will just add the rest.

To run Kedro-viz, first run the below command in your Databricks notebook: 

```bash
%sh kedro viz --no-browser --host 0.0.0.0 --port 4141
```

```eval_rst

> .. note:: The command execution continues to run and will need to be cancelled manually before proceeding to the next step. Cancelling the command will not quit the kedro-viz server.

Reviewer 2: What does cancelling manually mean? What would the user have to do?

Author: i have added a gif here which explains how to manually cancel

```

After this, you must try and run an example Shiny app using the below command: 

```bash
%r
library(shiny)
runExample("01_hello")

> ```

Reviewer 3: I am not a Databricks user, why do we need to call R's shiny app, and does that mean we also have dependencies on R?

![](../meta/images/databricks_shiny_command.png)

> When you click on the Shiny app link, it will open a browser with an example Shiny app running. Now edit the port at the end of the URL and change it to 4141. 

Author: can we pair-program on this on databricks as I am not sure how to get this working

> When you click on the Shiny app link, it will open a browser with an example Shiny app running. Now edit the port at the end of the URL and change it to 4141. 

Reviewer 4: I also thought this originally but unfortunately I don't think we can simplify this any more.

![](../meta/images/databricks_kedro_viz.gif)

You will notice Kedro-viz is blank at first, but if you click on the flowchart tab, Kedro-viz will run as normal. 

```eval_rst

> .. note:: Ctrl+C to quit the Kedro-viz server doesn't work on Databricks. To end the process you will need to find the process-id on port=4141 or the port you used and kill it using the Linux command kill -9 PID

Reviewer 2: ```suggestion
.. note:: Ctrl+C to quit the Kedro-viz server doesn't work on Databricks. To end the process you will need to find the process-id on port=4141 or the port you used and kill it using the Linux command `kill -9 PID`.
```

Reviewer 2: I don't know if my command highlight is correct here.

Reviewer: Is Ctrl+C what users would do both on Mac on Windows?

Author: yes!

Your complete notebook should look similar to this (the results are hidden):

![](../meta/images/databricks_notebook_example.png)

> ### 9. Using IPython Extension 

Reviewer: ```suggestion
### 9. Using the Kedro IPython Extension 
```

You can interact with Kedro in Databricks through the Kedro [IPython extension](https://ipython.readthedocs.io/en/stable/config/extensions/index.html), `kedro.extras.extensions.ipython`.

The Kedro IPython extension launches a [Kedro session](../kedro_project_setup/session.md) and makes available the useful Kedro variables `catalog`, `context`, `pipelines` and `session`. It also provides the `%reload_kedro` [line magic](https://ipython.readthedocs.io/en/stable/interactive/magics.html) that reloads these variables (for example, if you need to update `catalog` following changes to your Data Catalog).

> The IPython extension can be used in a Databricks notebook in a similar way to how it is used in [Jupyter notebooks](https://kedro.readthedocs.io/en/0.18.0/tools_integration/ipython.html).

Reviewer 4: Links to other docs pages from inside the docs should be relative so that they always point to the right version when build on RTD:
```suggestion
The IPython extension can be used in a Databricks notebook in a similar way to how it is used in [Jupyter notebooks](../tools_integration/ipython.md).
```

> Currently in Databricks, when you try to do a kedro run, it throws you a `ContextualVersionConflictError` because it downgrades `pip` to 21.0.1. We are not sure why this happens and hence there's one additional step you need to do in the Databricks notebook to make use of the IPython extension. After you load the IPython extension using the below command:

Reviewer 4: This makes us sound too incompetent, even if it's true 😝 
```suggestion
If you encounter a `ContextualVersionConflictError`, it is likely caused by Databricks using an old version of `pip`. Hence there's one additional step you need to do in the Databricks notebook to make use of the IPython extension. After you load the IPython extension using the below command:
```

```ipython

> In [1]: %load_ext kedro.extras.extensions.ipython

Reviewer 3: I need to do this before loading the extension. As my current directory is not in the project directory.
`%cd $project_root`

Author: thanks, I will also add this step

```

> You must explicitly upgrade your `pip` version by doing the below so it is on the latest version `pip ~= 22.0.4`. 

Reviewer 4: Let's leave out the specific version number so we don't have to keep updating this...
```suggestion
You must explicitly upgrade your `pip` version by doing the below: 
```

```bash
%pip install -U pip
```

After this, you can reload Kedro by running the line magic command `%reload_kedro <path_to_project_root>`.

### 10. Running Kedro-Viz on Databricks 

> Currently, if you try to run `%run_viz` on Databricks it will only display the below instead of running the Kedro-Viz app in the notebook. 

Reviewer 3: `%run_viz` isn't a command available. I got this error, do I need to load it manually?

`UsageError: Line magic function `%run_viz` not found.`

Author: Hi Nok, this is what i thought would happened. It's because kedro, and kedro-viz are installed in your databricks seperately. You will either have to uninstall kedro from the Cluster Library like i showed you and then do pip i kedro. Or install kedro-viz on the cluster library. if you do tht, run_viz should work. I will also highlight this in the docs!!!

Reviewer 4: Definitely worth noting this I think because it's not at all obvious. Something like this:
> Make sure Kedro-Viz and Kedro are both installed in the same scope (notebook-scoped vs. cluster library). i.e. if you `%pip install kedro` from inside your notebook then you should also `%pip install kedro-viz` from inside your notebook. If your cluster comes with Kedro installed on it as a library already then you should also [add Kedro-Viz as a cluster library](https://docs.microsoft.com/en-us/azure/databricks/libraries/cluster-libraries).

Author: yes, will add this

```console
<IPython.core.display.HTML object>
```

> While we fix this issue,  we have a temporary workaround which involves you setting up a R Shiny application to run Kedro-Viz on Databricks.

Reviewer: ```suggestion
While we fix this issue, we have a temporary workaround which involves you setting up a R Shiny application to run Kedro-Viz on Databricks.
```

> To run Kedro-Viz, first run the below command in your Databricks notebook: 

Reviewer 4: ```suggestion
To run Kedro-Viz, first ensure that you are in your Kedro project directory and then run the below command in your Databricks notebook: 
```

```bash
%sh kedro viz --no-browser --host 0.0.0.0 --port 4141
```

```eval_rst

> .. note:: The command execution continues to run and will need to be cancelled manually before proceeding to the next step. Cancelling the command will not quit the Kedro-Viz server. Please see below gif on how to cancel

Reviewer: ```suggestion
.. note:: The command execution continues to run and will need to be cancelled manually before proceeding to the next step. Cancelling the command will not quit the Kedro-Viz server. Please see below gif on how to cancel.
```
